src/java_bytecode/java_static_initializers.cpp

```diff
@@ -0,0 +1,186 @@
+  const irep_idt &class_name, const symbol_tablet &symbol_table)
+{
+  if(symbol_table.has_symbol(clinit_function_name(class_name)))
+    return true;
+
+  const class_typet &class_type =
+    to_class_type(symbol_table.lookup_ref(class_name).type);
+
+  for(const class_typet::baset &base : class_type.bases())
+  {
+    if(symbol_table.has_symbol(
+         clinit_wrapper_name(to_symbol_type(base.type()).get_identifier())))
+    {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/// Creates a static initializer wrapper for the given class. This function will
+/// call static initializer wrappers for all parent classes that have them, then
+/// run this class' clinit method if one exists.
+/// \param class_name: class symbol name
+/// \param symbol_table: global symbol table; will gain a clinit_wrapper symbol
+///   and a corresponding has-run-already global.
+static void create_clinit_wrapper(
+  const irep_idt &class_name, symbol_tablet &symbol_table)
+{
+  // Otherwise, assume that class C extends class C' and implements interfaces
+  // I1, ..., In. We now create the following function (possibly recursively
+  // creating the clinit_wrapper functions for C' and I1, ..., In):
+  //
+  // java::C::clinit_wrapper()
+  // {
+  //   if (java::C::clinit_already_run == false)
+  //   {
+  //     java::C::clinit_already_run = true; // before recursive calls!
+  //
+  //     java::C'::clinit_wrapper();
+  //     java::I1::clinit_wrapper();
+  //     java::I2::clinit_wrapper();
+  //     // ...
+  //     java::In::clinit_wrapper();
+  //
+  //     java::C::<clinit>();
+  //   }
+  // }
+  const irep_idt &already_run_name =
+    clinit_already_run_variable_name(class_name);
+  symbolt already_run_symbol;
+  already_run_symbol.name = already_run_name;
+  already_run_symbol.pretty_name = already_run_name;
+  already_run_symbol.base_name = "clinit_already_run";
+  already_run_symbol.type = bool_typet();
+  already_run_symbol.value = false_exprt();
+  already_run_symbol.is_lvalue = true;
+  already_run_symbol.is_state_var = true;
+  already_run_symbol.is_static_lifetime = true;
+  already_run_symbol.mode = ID_java;
+  symbol_table.add(already_run_symbol);
+
+  equal_exprt check_already_run(
+    already_run_symbol.symbol_expr(),
+    false_exprt());
+
+  // the entire body of the function is an if-then-else
+  code_ifthenelset wrapper_body;
+
+  // add the condition to the if
+  wrapper_body.cond()=check_already_run;
+
+  // add the "already-run = false" statement
+  code_blockt init_body;
+  code_assignt set_already_run(already_run_symbol.symbol_expr(), true_exprt());
+  init_body.move_to_operands(set_already_run);
+
+  // iterate through the base types and add recursive calls to the
+  // clinit_wrapper()
+  const symbolt &class_symbol = *symbol_table.lookup(class_name);
+  for(const auto &base : to_class_type(class_symbol.type).bases())
+  {
+    const auto base_name = to_symbol_type(base.type()).get_identifier();
```

Comment:
@svorenova - you should have a look at this and ensure your change for generic base classes won't interfere with this (And vice versa)

Reply:
While the PR for my changes is still awaiting approval, I'm fairly confident at this point that there should be no interference. I will keep this in mind though until the PR is merged.